bitcoind: Add -daemonwait option to wait for initialization

laanwj opened the pull request:
This adds a `-daemonwait` flag that does the same as `-daemon` except that it, from a user perspective, backgrounds the process only after initialization is complete. This is similar to the behaviour of some other software such as c-lightning.

This can be useful when the process launching bitcoind wants to guarantee that either the RPC server is running, or that initialization failed, before continuing. The exit code indicates the initialization result.

The use of the libc function `daemon()` is replaced by a custom implementation which is inspired by the [glibc implementation](https://github.com/lattera/glibc/blob/master/misc/daemon.c#L44), but which also creates a pipe from the child to the parent process for communication.

An additional advantage of having our own `daemon()` implementation is that no MACOS-specific pragmas are needed anymore to silence a deprecation warning.

TODO:

- [x] Factor out `token_read` and `token_write` to an utility, and use  them in `shutdown.cpp` as well—this is exactly the same kind of communication mechanism.

    - [x] RAII-ify pipe endpoints.

- [x] Improve granularity of the `configure.ac` checks. This currently  still checks for the function `daemon()` which makes no sense as  it's not used. It should check for individual functions such as
  `fork()` and `setsid()` etc—the former being required, the second optional.

- [-] ~~Signal propagation during initialization: if say, pressing Ctrl-C during `-daemonwait` it would be good to pass this SIGINT on to the child process instead of detaching the parent process and letting the child run free.~~ This is not necessary, see https://github.com/bitcoin/bitcoin/pull/21007#issuecomment-769007341.

Future:

- Consider if it makes sense to use this in the RPC tests (there would be no more need for "is RPC ready" polling loops). I think this is out of scope for this PR.

laanwj committed:
bitcoind: Add -daemonwait option to wait for initialization

This adds a `-daemonwait` flag that does the same as `-daemon` except
it, from a user perspective, backgrounds the process only after
initialization is complete.

This can be useful when the process launching bitcoind wants to
guarantee that either the RPC server is running, or that initialization
failed, before continuing. The exit code indicates the initialization
result.

This replaces the use of the libc function `daemon()` by a custom
implementation which is inspired by the glibc implementation, but also
creates a pipe from the child to the parent process for communication.

An additional advantage of having our own `daemon()` implementation is
that no MACOS-specific pragmas are needed anymore to silence a
deprecation warning.

hebasto commented:
> This can be useful when the process launching bitcoind wants to guarantee that either the RPC server is running, or that initialization failed, before continuing. The exit code indicates the initialization result.

Why such behavior couldn't be the default?

hebasto commented on src/init.cpp:
```diff
@@ -577,7 +577,8 @@ void SetupServerArgs(NodeContext& node)
 
 #if HAVE_DECL_DAEMON
-    argsman.AddArg("-daemon", "Run in the background as a daemon and accept commands", ArgsManager::ALLOW_ANY, OptionsCategory::OPTIONS);
+    argsman.AddArg("-daemon", "Run in the background as a daemon and accept commands (default: 0)", ArgsManager::ALLOW_ANY, OptionsCategory::OPTIONS);
+    argsman.AddArg("-daemonwait", "Wait for initialization to be finished before exiting. This implies -daemon (default: 0)", ArgsManager::ALLOW_ANY, OptionsCategory::OPTIONS);
```
s/`ALLOW_ANY`/`ALLOW_BOOL`/ in both lines?

laanwj replied:
Good idea!
I'm also surprised there are no DEFAULT_XXX constants here.

laanwj commented:
> Why such behavior couldn't be the default?

I didn't do that to avoid that discussion for now :slightly_smiling_face: . It could always be made default at some point in the future if that's what people want.

hebasto commented:
Concept ACK.

practicalswift commented:
Concept ACK

laanwj committed:
shutdown: Use RAII TokenPipe in shutdown

laanwj committed:
util: Add RAII TokenPipe

theStack commented:
Concept ACK

promag reviewed (commented):
Tested ACK 7482634cd455c6d25287e52795c2ac78cb1d44b0 on macos.

With the following diff
```diff
--- a/src/init.cpp
+++ b/src/init.cpp
@@ -780,6 +780,9 @@ static bool InitSanityCheck()

 static bool AppInitServers(const util::Ref& context, NodeContext& node)
 {
+    sleep(5);
+    return false;
+
     const ArgsManager& args = *Assert(node.args);
     RPCServer::OnStarted(&OnRPCStarted);
     RPCServer::OnStopped(&OnRPCStopped);
```
On master
```
./src/bitcoind -regtest -daemon  && echo running
Bitcoin Core starting
running
```
But it's not running :trollface: 

And with this change:
```
./src/bitcoind -regtest -daemonwait && echo running
Bitcoin Core starting
Error during initializaton - check debug.log for details
```

Good stuff, still have to review code change.

laanwj commented:
Thanks for testing! FWIW what I use to force a failure in `AppInitMain` is
```
src/bitcoind -regtest -daemonwait -pid=/invalid/path
```
(pid is, by definition, only written in the daemon process)
But that works too :smile:

hebasto commented:
> (pid is, by definition, only written in the daemon process)

Not sure if this is true in the master branch.

laanwj commented:
> Not sure if this is true in the master branch.

I am sure of this. I did not change this at all, and besides, it would be a bug otherwise. After all, what use is the temporary parent process' PID.

hebasto commented:
> ... it would be a bug otherwise.

```
src/bitcoind -daemon=0
```
creates `bitcoind.pid`

laanwj commented:
> creates bitcoind.pid

That's not my point. My point is that it creates the PID file *after* daemonizing. Sure, it also creates the PID file otherwise, that's okay.

jonatack reviewed (commented):
Tested at 7482634cd455c6d25287e52795c2ac78cb1d44b0, debug-builds cleanly for me and seems to behave as expected.

invalid path
```
$ ./src/bitcoind -signet -daemonwait -pid=/invalid/path
Bitcoin Core starting
Error during initializaton - check debug.log for details
```

(this particular case does not log anything to the debug log)


happy path
```
$ ./src/bitcoind -signet -daemonwait ; ./src/bitcoin-cli -signet stop
Bitcoin Core starting
Bitcoin Core stopping
```
debug log is as expected
```
2021-01-27T17:49:33Z Command-line arg: daemonwait=""
2021-01-27T17:49:33Z Command-line arg: signet=""
2021-01-27T17:49:33Z Using at most 125 automatic connections (1024 file descriptors available)
...
2021-01-27T17:49:35Z init message: Done loading
2021-01-27T17:49:35Z dnsseed thread start
2021-01-27T17:49:35Z Waiting 11 seconds before querying DNS seeds.
2021-01-27T17:49:35Z tor: Thread interrupt
2021-01-27T17:49:35Z addcon thread exit
...

2021-01-27T17:49:36Z Shutdown: done
```
passing `-daemonwait=1` behaves (and logs) as expected as well

Ctrl-C works with a second or two of delay; it seems to not take effect until the wait is released but I'm not sure.

laanwj commented:
Thanks for testing @jonatack ,

> (this particular case does not log anything to the debug log)

That it doesn't log anything is peculiar ! (but unrelated to this PR, I guess, the same problem would happen with `-daemon`, it would just mysteriously exit in the background).

> Ctrl-C works with a second or two of delay; it seems to not take effect until the wait is released but I'm not sure.

It does respond to it? That's interesting. I thought the child process wouldn't get the SIGINT signal at all (hence my remaining TODO). That it doesn't react immediately is normal, It can't interrupt during verification of a block, for example. it's the same when interrupting a non-daemon startup.

jonatack commented:
Tested with testnet, which is much slower than signet for me (95 sec vs 3)
```
$ time ./src/bitcoind -testnet -daemonwait=1 && ./src/bitcoin-cli -testnet stop
Bitcoin Core starting

real	1m35.265s
user	0m0.137s
sys	0m0.015s
Bitcoin Core stopping
$
```

Ctrl-C right after launching interrupts successfully after the same time period elapses 
```
$ time ./src/bitcoind -testnet -daemonwait=1
Bitcoin Core starting
^C
real	1m39.239s
user	0m0.130s
sys	0m0.022s
$
```

laanwj commented:
I've looked into this a bit.

It looks like that until the parent process exits, the child process is still attached to the terminal, so it still gets the `SIGINT` from Ctrl-C directly. Even though it did `dup2` `stdin/stdout/stderr` to `/dev/null` already.

So there appears to be no need to do explicit signal propagation. That's neat. I'll remove that TODO and open this for review.

theStack commented on src/util/tokenpipe.h:
```diff
@@ -0,0 +1,98 @@
+        TS_EOS = -2, //!< Unexpected end of stream
+    };
+
+    /** Read token from file descriptor.
```
```suggestion
    /** Write token to file descriptor.
```

theStack commented on src/bitcoind.cpp:
```diff
@@ -105,24 +187,34 @@ static bool AppInit(int argc, char* argv[])
+                case 0: // Child: continue.
+                    // If -daemonwait is not enabled, immediately send a success token the parent.
+                    if (!args.GetBoolArg("-daemonwait", DEFAULT_DAEMONWAIT)) {
+                        daemon_ep.TokenWrite(true);
```
nit, feel free to ignore: maybe avoid implicit casting from bool to uint8_t and use literal 1 (or some character literal) as success token instead? I think as it is this could be potentially confusing for readers (also the counterpart `TokenRead` call below). As a drawback though, `-daemonwait` code gets longer: `daemon_ep.TokenWrite(fRet ? 1 : 0)`.

laanwj replied:
Agree, it's probably better to pass some constant.

theStack reviewed (commented):
Nice idea of introducing a RAII TokenPipe! This reduces the amount of low-level system calls / error handling code in the bitcoind module, looks much cleaner now.
Left some review comments below, mostly nitpicking. Will review the the third commit (i.e. the actual introduction of `-daemonwait`) more in detail and also to test it on the weekend.

laanwj commented on src/util/tokenpipe.cpp:
```diff
@@ -0,0 +1,98 @@
+{
+#if HAVE_O_CLOEXEC
+    if (pipe2(m_fds, O_CLOEXEC) != 0) {
+        return;
```
I was first thinking of using the rust trick: to have a private constructor, and a static public factory function on the class that returns an `std::optional<Pipe>`. But it would complicate some things. E.g. currently this type is not movable at all, it felt like overkill to implement that just for initialization, but maybe it's worth it if we can make it a standard idiom.

laanwj commented on src/util/tokenpipe.cpp:
```diff
@@ -0,0 +1,98 @@
+
+TokenPipeEnd::~TokenPipeEnd()
+{
+    if (m_fd != -1) close(m_fd);
```
Yes, why not. It seemed like so little code that it wouldn't really be worth deduplicating. I agree the redundant set to `-1` doesn't matter, this is not a performance critical function.

laanwj commented:
Addressed @theStack's comments:
- Added `TokenPipe::create()` that returns an optional (and make `TokenPipe` movable), instead of the awkward incomplete construction.
- Call `Close()` in destructor instead of duplicating code.
- Consistently pass int to `TokenWrite`.
- Fixed typos in comments.

theStack commented on src/shutdown.cpp:
```diff
@@ -24,25 +27,17 @@ std::mutex g_shutdown_mutex;
-        return false;
-    }
-#endif
+    std::optional<TokenPipe> pipe = TokenPipe::create();
+    g_shutdown_r = pipe->ReadEnd();
+    g_shutdown_w = pipe->WriteEnd();
```
The error-check after TokenPipe creation is missing here, e.g. `if (!pipe) return false;`

laanwj replied:
Good point.

theStack commented on src/bitcoind.cpp:
```diff
@@ -28,6 +29,79 @@
+    // communication pipe with child process
+    std::optional<TokenPipe> umblical = TokenPipe::create();
+    if (!umblical) {
+        return -1; // pipe or pip2 failed.
```
```suggestion
        return -1; // pipe or pipe2 failed.
```

theStack commented on src/bitcoind.cpp:
```diff
@@ -105,24 +187,34 @@ static bool AppInit(int argc, char* argv[])
+                    return InitError(Untranslated(strprintf("fork_daemon() failed: %s\n", strerror(errno))));
+                default: { // Parent: wait and exit.
+                        int token = daemon_ep.TokenRead();
+                        if (token == true) { // fRet == true, Success
```
nit:
```suggestion
                        if (token == 1) { // fRet == 1, Success
```
(probably the token variable could be eliminated by calling `TokenRead` directly in the if condition.)

laanwj replied:
I prefer having a variable here. Having a read as a side effect of an if() clause is kind of meh.
Will change the `true`.

theStack commented on src/bitcoind.cpp:
```diff
@@ -59,6 +133,14 @@ static bool AppInit(int argc, char* argv[])
+#if HAVE_DECL_FORK
+    // Communication with parent after daemonizing. This is used for signalling in the following ways:
+    // - a boolean token is sent when the initialization process (all the Init* functions) have finished to indicate
+    // that the parent process can quit, and whether it was succesful/unsuccesful.
```
typos:
```suggestion
    // that the parent process can quit, and whether it was successful/unsuccessful.
```

theStack commented on src/bitcoind.cpp:
```diff
@@ -28,6 +29,79 @@
+    }
+    if (pid != 0) {
+        // Parent process gets read end, closes write end.
+        endpoint = umblical->ReadEnd();
```
I didn't know the word before (nice name for a parent-child-pipe 😄), but according to the dictionary the spelling is slightly different:
```suggestion
        endpoint = umbilical->ReadEnd();
```

theStack reviewed (commented):
Finished reviewing the code, LGTM. Left some comments, mostly nits and typos again.

laanwj commented:
Thanks again for the review @theStack 
Pushed https://github.com/bitcoin/bitcoin/compare/48858432bc7ef3e02685b693954173e722bae4e5..652d4e4382e8964db3e1ba6e4bb8247daad7ae64

promag reviewed (commented):
Tested ACK 652d4e4382.

Should disallow conflict options like `-daemon=false -daemonwait=true`?

laanwj commented:
Thanks for testing!

> Should disallow conflict options like -daemon=false -daemonwait=true?

I'm not sure they're really conflicting ("enable wait-for-daemon, but don't daemonize"), though pointless, I see no reason to add a special error message for this.

promag commented:
My point is with `-daemon=false -daemonwait=true` it will daemonize.

laanwj commented:
> My point is with -daemon=false -daemonwait=true it will daemonize.

That's intentional, it means that you can just write `bitcoind -daemonwait` and it will work.

laanwj commented:
Rebased for  #21250.

theStack reviewed (approved):
Tested ACK 8e6d339434fdb6cfe5057b60f89ce08101e2eeb0 :beers:
(ran on Debian GNU/Linux bullseye/sid, Kernel 4.19.0)

ajtowns commented on src/bitcoind.cpp:
```diff
@@ -28,6 +29,79 @@
+    }
+    // Child process gets write end, closes read end.
+    endpoint = umbilical->WriteEnd();
+    umbilical->ReadEnd();
```
Calling `umbilical->ReadEnd()` to create the `TokenPipeEnd` then immediately destroy it, which then closes the underlying fd seems a bit clunky. A comment might be worthwhile at least.

laanwj replied:
I thought about adding an explicit `CloseReadEnd` `CloseWriteEnd` function but as it'd essentially be the same I didn't do so. Will add a comment.

laanwj replied:
What if I call the functions `TakeReadEnd` `TakeWriteEnd` would that make it clearer that a move is happening?

ajtowns replied:
Could also write `umbilical->ReadEnd().Close();` (relying on the destructor's second call to Close then being a no-op). I don't really have a preference.

laanwj replied:
Ah yea! I think I'll do both.
```
umbilical->TakeReadEnd().Close();
```
Fully self-documenting then.

ajtowns commented on src/util/tokenpipe.cpp:
```diff
@@ -0,0 +1,108 @@
+void TokenPipe::Close()
+{
+    if (m_fds[0] != -1) close(m_fds[0]);
+    if (m_fds[1] != -1) close(m_fds[1]);
```
I think this would be incorrect if you had code like: `{ TokenPipe tp; a(); tp.Close(); b(); }` in particular -- the fds would be closed when `Close()` is called, but may be reallocated to different files during `b()` which would then be incorrectly closed when `tp` is destructed.

Could explicitly set `m_fds = {-1,-1}`, or could drop `Close()` entirely and move the code into the destructor -- it's only called from `operator=(TokenPipe&&)` and you could just mark that as `= deleted;` I think; it doesn't seem to be needed.

laanwj replied:
Will take a look. Agree the fds should be se to -1 after closing them to prevent 'use fd after close' problems, seems I forgot that. I made it like this to be symmetric with `TokenPipeEnd::Close`. Sure, I could move the code around if minimalism was the goal, but I like a consistent API more.

ajtowns commented on src/util/tokenpipe.cpp:
```diff
@@ -0,0 +1,108 @@
+#include <fcntl.h>
+#include <unistd.h>
+
+std::optional<TokenPipe> TokenPipe::create()
```
Doesn't matter, but net_processing uses `make` as the name for PeerManager's nearly-but-not-quite-a-constructor.

laanwj replied:
Thanks, I'm fine with a different name, `create` is the convention in rust but Make is fine with me too.

ajtowns replied:
Yeah, either way. PeerManager could change too; it was just copied from some stackoverflow link or something.

ajtowns reviewed (commented):
utACK 8e6d339434fdb6cfe5057b60f89ce08101e2eeb0

jonatack commented on src/Makefile.am:
```diff
@@ -240,6 +240,7 @@ BITCOIN_CORE_H = \
   util/memory.h \
   util/message.h \
   util/moneystr.h \
+  util/tokenpipe.h \
```
nit, sort

laanwj replied:
Strange, I guess I had this module named differently at first :smile:

jonatack commented on src/util/tokenpipe.cpp:
```diff
@@ -0,0 +1,108 @@
+int TokenPipeEnd::TokenWrite(uint8_t token)
+{
+    while (true) {
+        int result = write(m_fd, &token, 1);
```
`write` seems to be returning ssize_t / long, same for `read` line 86
```suggestion
        const long result{write(m_fd, &token, 1)};
```

laanwj replied:
good catch with the type, not sure though i like the `{}` initialization style more here, `=` seems to express intent clearer

jonatack commented on src/init.h:
```diff
@@ -9,6 +9,11 @@
+//! Default value for -daemon option
+static const bool DEFAULT_DAEMON = false;
+//! Default value for -daemonwait option
+static const bool DEFAULT_DAEMONWAIT = false;
```
these two can be `constexpr`

jonatack commented on src/bitcoind.cpp:
```diff
@@ -28,6 +29,79 @@
+        // from terminal.
+        int fd = open("/dev/null", O_RDWR);
+        if (fd >= 0) {
+            bool err = dup2(fd, STDIN_FILENO) < 0 || dup2(fd, STDOUT_FILENO) < 0 || dup2(fd, STDERR_FILENO) < 0;
```
`fd` and `err` can be const

laanwj replied:
I'm not really sure what the common thing is to do for internal variables, do we have a const-unless-required-otherwise recommendation?

jonatack replied:
Yes, I don't think it's explicitly stated, other than maybe via the C++ Code Guidelines linked to in the developer notes, but it's what I've been doing/reviewing, and seeing others say and do.

jonatack commented on src/shutdown.cpp:
```diff
@@ -96,17 +81,10 @@ void WaitForShutdown()
-            assert(result == 1);
-            break;
-        }
+    int res = g_shutdown_r.TokenRead();
```
`res` here and line 54 can be const

jonatack reviewed (commented):
tACK 8e6d339434fdb6cfe5057b60f89ce08101e2eeb0

Some nit suggestions below, feel free to ignore.

hebasto commented on src/util/tokenpipe.cpp:
```diff
@@ -0,0 +1,108 @@
+    return res;
+}
+
+TokenPipeEnd::TokenPipeEnd(int fd):
+    m_fd(fd)
```
style-nit, suggested by [clang-format-diff.py](https://github.com/bitcoin/bitcoin/tree/master/contrib/devtools#clang-format-diffpy):
```suggestion
TokenPipeEnd::TokenPipeEnd(int fd)
    : m_fd(fd)
```

hebasto commented on src/util/tokenpipe.h:
```diff
@@ -0,0 +1,123 @@
+#include <optional>
+
+/** One end of a token pipe. */
+class TokenPipeEnd {
+private:
+    int m_fd = -1;
+public:
```
style-nit, suggested by [clang-format-diff.py](https://github.com/bitcoin/bitcoin/tree/master/contrib/devtools#clang-format-diffpy):
```suggestion
class TokenPipeEnd
{
private:
    int m_fd = -1;

public:
```

hebasto commented on src/util/tokenpipe.h:
```diff
@@ -0,0 +1,123 @@
+        other.m_fd = -1;
+        return *this;
+    }
+    TokenPipeEnd (const TokenPipeEnd&) = delete;
+    TokenPipeEnd &operator=(const TokenPipeEnd&) = delete;
```
style-nit, suggested by [clang-format-diff.py](https://github.com/bitcoin/bitcoin/tree/master/contrib/devtools#clang-format-diffpy):
```suggestion
    TokenPipeEnd(const TokenPipeEnd&) = delete;
    TokenPipeEnd& operator=(const TokenPipeEnd&) = delete;
```

hebasto commented on src/util/tokenpipe.h:
```diff
@@ -0,0 +1,123 @@
+/** An interprocess or interthread pipe for sending tokens (one-byte values)
+ * over.
+ */
+class TokenPipe {
```
style-nit, suggested by [clang-format-diff.py](https://github.com/bitcoin/bitcoin/tree/master/contrib/devtools#clang-format-diffpy):
```suggestion
class TokenPipe
{
```

hebasto commented on src/util/tokenpipe.h:
```diff
@@ -0,0 +1,123 @@
+private:
+    int m_fds[2] = {-1, -1};
+
+    TokenPipe(int fds[2]): m_fds{fds[0], fds[1]} {}
+public:
```
style-nit, suggested by [clang-format-diff.py](https://github.com/bitcoin/bitcoin/tree/master/contrib/devtools#clang-format-diffpy):
```suggestion
    TokenPipe(int fds[2]) : m_fds{fds[0], fds[1]} {}

public:
```

hebasto commented on src/util/tokenpipe.h:
```diff
@@ -0,0 +1,123 @@
+    // Move-only class.
+    TokenPipe(TokenPipe&& other)
+    {
+        for(int i = 0; i < 2; ++i) {
```
style-nit, suggested by [clang-format-diff.py](https://github.com/bitcoin/bitcoin/tree/master/contrib/devtools#clang-format-diffpy):
```suggestion
        for (int i = 0; i < 2; ++i) {
```

hebasto commented on src/util/tokenpipe.h:
```diff
@@ -0,0 +1,123 @@
+    TokenPipe& operator=(TokenPipe&& other)
+    {
+        Close();
+        for(int i = 0; i < 2; ++i) {
```
style-nit, suggested by [clang-format-diff.py](https://github.com/bitcoin/bitcoin/tree/master/contrib/devtools#clang-format-diffpy):
```suggestion
        for (int i = 0; i < 2; ++i) {
```

hebasto commented on src/util/tokenpipe.h:
```diff
@@ -0,0 +1,123 @@
+        }
+        return *this;
+    }
+    TokenPipe (const TokenPipe&) = delete;
+    TokenPipe &operator=(const TokenPipe&) = delete;
```
style-nit, suggested by [clang-format-diff.py](https://github.com/bitcoin/bitcoin/tree/master/contrib/devtools#clang-format-diffpy):
```suggestion
    TokenPipe(const TokenPipe&) = delete;
    TokenPipe& operator=(const TokenPipe&) = delete;
```

laanwj replied:
Thanks. FWIW it's fine to just mark one of these and say to run `clang-format-diff`, no need to open a review item for every little indentation/style mismatch :smile:

hebasto commented on src/bitcoind.cpp:
```diff
@@ -28,6 +29,79 @@
+ *          In case of success, endpoint will be one end of a pipe from the child to parent process,
+ *          which can be used with TokenWrite (in the child) or TokenRead (in the parent).
+ */
+int fork_daemon(bool nochdir, bool noclose, TokenPipeEnd &endpoint)
```
style-nit, suggested by [clang-format-diff.py](https://github.com/bitcoin/bitcoin/tree/master/contrib/devtools#clang-format-diffpy):
```suggestion
int fork_daemon(bool nochdir, bool noclose, TokenPipeEnd& endpoint)
```

hebasto commented on src/bitcoind.cpp:
```diff
@@ -105,24 +187,34 @@ static bool AppInit(int argc, char* argv[])
             // Daemonize
-            if (daemon(1, 0)) { // don't chdir (1), do close FDs (0)
-                return InitError(Untranslated(strprintf("daemon() failed: %s\n", strerror(errno))));
+            switch (fork_daemon(1, 0, daemon_ep)) { // don't chdir (1), do close FDs (0)
+                case 0: // Child: continue.
+                    // If -daemonwait is not enabled, immediately send a success token the parent.
+                    if (!args.GetBoolArg("-daemonwait", DEFAULT_DAEMONWAIT)) {
+                        daemon_ep.TokenWrite(1);
+                        daemon_ep.Close();
+                    }
+                    break;
+                case -1: // Error happened.
+                    return InitError(Untranslated(strprintf("fork_daemon() failed: %s\n", strerror(errno))));
+                default: { // Parent: wait and exit.
+                        int token = daemon_ep.TokenRead();
+                        if (token) { // Success
+                            exit(EXIT_SUCCESS);
+                        } else { // fRet = false or token read error (premature exit).
+                            tfm::format(std::cout, "Error during initializaton - check debug.log for details\n");
+                            exit(EXIT_FAILURE);
+                        }
+                    }
```
indentation-nit, suggested by [clang-format-diff.py](https://github.com/bitcoin/bitcoin/tree/master/contrib/devtools#clang-format-diffpy):
```suggestion
            switch (fork_daemon(1, 0, daemon_ep)) { // don't chdir (1), do close FDs (0)
            case 0: // Child: continue.
                // If -daemonwait is not enabled, immediately send a success token the parent.
                if (!args.GetBoolArg("-daemonwait", DEFAULT_DAEMONWAIT)) {
                    daemon_ep.TokenWrite(1);
                    daemon_ep.Close();
                }
                break;
            case -1: // Error happened.
                return InitError(Untranslated(strprintf("fork_daemon() failed: %s\n", strerror(errno))));
            default: // Parent: wait and exit.
                int token = daemon_ep.TokenRead();
                if (token) { // Success
                    exit(EXIT_SUCCESS);
                } else { // fRet = false or token read error (premature exit).
                    tfm::format(std::cout, "Error during initializaton - check debug.log for details\n");
                    exit(EXIT_FAILURE);
               }
```

hebasto commented on src/util/tokenpipe.h:
```diff
@@ -0,0 +1,123 @@
+    /** Create a new pipe.
+     * @returns The created TokenPipe, or an empty std::nullopt in case of error.
+     */
+    static std::optional<TokenPipe> create();
```
Capitalize function name as others?
```suggestion
    static std::optional<TokenPipe> Create();
```

hebasto commented on src/util/tokenpipe.h:
```diff
@@ -0,0 +1,123 @@
+    TokenPipe &operator=(const TokenPipe&) = delete;
+};
+
+#endif
```
```suggestion
#endif // WIN32
```

hebasto reviewed (commented):
Approach ACK 8e6d339434fdb6cfe5057b60f89ce08101e2eeb0

ajtowns commented on src/bitcoind.cpp:
```diff
@@ -28,6 +29,79 @@
+            if (err) {
+                exit(1); // dup2 failed.
+            }
+            close(fd);
```
The standard reason to compare it to 2 is in case you start the executable with stdin/stdout/stderr already closed -- in that case `fd` will be assigned to 0/1/2 because that's the first free fd, `dup2(fd,stdin)` (or out or err) will evaluate to `dup2(0,0)` and be a no-op, and you'll then call `close(0)` which will leave you without a dummied stdin (or out or err) which will mean some later fd you open might overlap with stdin (or out or err). Not sure that things wouldn't already be broken in those cases though -- TokenPipe would probably grab the free fds in that case, and then dup2 would close the TokenPipe fd.

laanwj replied:
Hmm good point. I did not think about that case. It's really unconventional to start binaries with stdin/stdout/stderr closed (the graceful way would be to redirect them to `/dev/null` and I'm not sure we handle this in other places.

laanwj commented:
@hebasto @jonatack @ajtowns
Thanks for the reviews. I have addressed your comments.

jonatack commented on src/bitcoind.cpp:
```diff
@@ -105,24 +187,34 @@ static bool AppInit(int argc, char* argv[])
+                if (token) { // Success
+                    exit(EXIT_SUCCESS);
+                } else { // fRet = false or token read error (premature exit).
+                    tfm::format(std::cout, "Error during initializaton - check debug.log for details\n");
```
b83b386 s/cout/cerr/?

laanwj replied:
Yes, why not

jonatack commented on src/util/tokenpipe.cpp:
```diff
@@ -0,0 +1,108 @@
+    m_fd = -1;
+}
+
+#endif
```
0b9994ac
```suggestion
#endif // WIN32
```

jonatack commented on src/bitcoind.cpp:
```diff
@@ -28,6 +29,79 @@
+ * Opens a pipe to the child process to be able to wait for an event to occur.
+ *
+ * @returns 0 if succesful, and in child process.
+ *          >0 if succesful, and in parent process.
```
b83b386c lines 37 and 38: s/succesful/successful/

jonatack reviewed (commented):
ACK b83b386c6e03a57f93110dfefdbbf515e62b5753 reviewed diff/code, built/tested

Some nits if you retouch.

laanwj commented:
Thanks @jonatack, have addressed the comments and re-pushed
[b83b386c6e03a57f93110dfefdbbf515e62b5753..9a09a494e59b61d7a185da44f9c6736cb536c0c1](https://github.com/bitcoin/bitcoin/compare/b83b386c6e03a57f93110dfefdbbf515e62b5753..9a09a494e59b61d7a185da44f9c6736cb536c0c1)

jonatack commented:
re-ACK 9a09a494e59b61d7a185da44f9c6736cb536c0c1

laanwj commented:
Repushed for an ordering issue in `tokenpipe.cpp` reported out-of-band, the implementation file now has same class order as the header
[9a09a494e59b61d7a185da44f9c6736cb536c0c1..e017a913d0d78ef0766cf73586fe7a38488e1a26](https://github.com/bitcoin/bitcoin/compare/9a09a494e59b61d7a185da44f9c6736cb536c0c1..e017a913d0d78ef0766cf73586fe7a38488e1a26)

jonatack commented:
Tested ACK e017a913d0d78ef0766cf73586fe7a38488e1a26 checked change since previous review is move-only

```
$ time ./src/bitcoind -testnet -daemonwait=1 && ./src/bitcoin-cli -testnet stop
Bitcoin Core starting

real	1m4.095s
user	0m0.132s
sys	0m0.027s
Bitcoin Core stopping

$ time ./src/bitcoind -testnet -daemonwait=1
Bitcoin Core starting
^C 
real	1m5.311s
user	0m0.047s
sys	0m0.023s
```

fanquake commented:
This change was part of 22.0, removing "Needs release note".
